File: remote/src/main/java/com/kelsos/mbrc/interactors/TrackRatingInteractorImpl.java
package com.kelsos.mbrc.interactors;

import com.google.inject.Inject;
import com.kelsos.mbrc.dto.requests.RatingRequest;
import com.kelsos.mbrc.dto.track.Rating;
import com.kelsos.mbrc.services.api.TrackService;
import rx.Observable;

public class TrackRatingInteractorImpl implements TrackRatingInteractor {
  @Inject private TrackService service;

  @Override public Observable<Rating> getRating() {
    return service.getTrackRating();
  }

  @Override public Observable<Float> updateRating(float rating) {
    return service.updateRating(new RatingRequest().setRating(rating))
        .flatMap(baseResponse -> Observable.just(rating));
  }
}

File: mbrc-lib/src/main/kotlin/com/kelsos/mbrc/interactors/TrackRatingInteractorImpl.kt
package com.kelsos.mbrc.interactors

import com.google.inject.Inject
import com.kelsos.mbrc.dto.requests.RatingRequest
import com.kelsos.mbrc.dto.track.Rating
import com.kelsos.mbrc.services.api.TrackService
import rx.Observable
import rx.functions.Func1

class TrackRatingInteractorImpl : TrackRatingInteractor {
    @Inject private lateinit var service: TrackService

    override fun getRating(): Observable<Rating> {
        return service.getTrackRating()
    }

    override fun updateRating(rating: Float): Observable<Float> {
        return service.updateRating(RatingRequest().setRating(rating))
                .flatMap<Float>(Func1 {
                    Observable.just(rating)
                })
    }
}
